server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java

```diff
@@ -0,0 +1,570 @@
+                        final OnDiskState onDiskState = loadOnDiskState(directoryReader);
+
+                        if (nodeEnvironment.nodeId().equals(onDiskState.nodeId) == false) {
+                            throw new IllegalStateException("unexpected node ID in metadata, found [" + onDiskState.nodeId +
+                                "] but expected [" + nodeEnvironment.nodeId() + "]");
+                        }
+
+                        if (onDiskState.metaData.clusterUUIDCommitted()) {
+                            if (committedClusterUuid == null) {
+                                committedClusterUuid = onDiskState.metaData.clusterUUID();
+                            } else if (committedClusterUuid.equals(onDiskState.metaData.clusterUUID()) == false) {
+                                throw new IllegalStateException("mismatched cluster UUIDs in metadata, found [" + committedClusterUuid +
+                                    "] and [" + onDiskState.metaData.clusterUUID() +
+                                    "]");
+                            }
+                        }
+
+                        maxCurrentTerm = Math.max(maxCurrentTerm, onDiskState.currentTerm);
+
+                        long acceptedTerm = onDiskState.metaData.coordinationMetaData().term();
+                        long maxAcceptedTerm = bestOnDiskState.metaData.coordinationMetaData().term();
+                        if (acceptedTerm > maxAcceptedTerm
+                            || (acceptedTerm == maxAcceptedTerm
+                                && (onDiskState.lastAcceptedVersion > bestOnDiskState.lastAcceptedVersion
+                                    || (onDiskState.lastAcceptedVersion == bestOnDiskState.lastAcceptedVersion)
+                                        && onDiskState.currentTerm > bestOnDiskState.currentTerm))) {
+                            bestOnDiskState = onDiskState;
+                        }
+                    } catch (IndexNotFoundException e) {
+                        logger.debug(new ParameterizedMessage("no on-disk state at {}", indexPath), e);
+                    }
+                }
+            }
+        }
+
+        if (bestOnDiskState.currentTerm != maxCurrentTerm) {
+            throw new IllegalStateException("inconsistent terms found: best state is in term [" + bestOnDiskState.currentTerm +
+                "] but there is a stale state with greater term [" + maxCurrentTerm + "]");
+        }
+
+        return bestOnDiskState;
+    }
+
+    private static Path getMetaDataIndexPath(Path path, int majorVersion) {
+        return path.resolve(getMetaDataIndexDirectoryName(majorVersion));
+    }
+
+    public static String getMetaDataIndexDirectoryName(int majorVersion) {
+        // include the version in the directory name to create a completely new index when upgrading to the next major version.
+        return "_metadata_v" + majorVersion;
```

Comment:
@jpountz you mentioned that there may be a better way to deal with upgrades than creating a completely new index for each major version. Could you share some pointers?

Reply:
Did I? A different index per major version sounds like the right approach to me.

Reply:
I think we can reuse the same directory, which will simplify things. To upgrade, we can just create a new `IndexWriter` with `OpenMode.CREATE` and use that one to write out the upgraded content, and finally commit. The IndexWriter will carry over the generation numbers, and version/counter, but otherwise create a new vanilla index (with new lucene created version). The benefit of this approach is that it solves atomic upgrades on the folder.

Update: This is actually how the implementation already works on "open", so it looks like we can just get rid of the versioned folders?

Reply:
Ok, I ditched the versioned folder in 492b3a73334889a78095290b13850286c71c5bf4. It's not clear that Lucene will always continue to offer us the guarantees we need for this workflow but it does today, and if it breaks in future we will find a way to deal with it.